infra/Pos/Network/CLI.hs

```diff
@@ -223,22 +226,35 @@ intNetworkConfigOpts cfg@NetworkConfigOpts{..} = do
                         Just fp -> liftIO $ readTopology fp
     ourTopology <- case parsedTopology of
       Y.TopologyStatic{..} -> do
-        (md@NodeMetadata{..}, initPeers) <- liftIO $ fromPovOf cfg topologyAllPeers
+        (md@NodeMetadata{..}, initPeers, kademliaPeers) <- liftIO $ fromPovOf cfg topologyAllPeers
         topologyStaticPeers <- monitorStaticConfig cfg md initPeers
+        -- If kademlia is enabled here then we'll try to read the configuration
+        -- file. However it's not necessary that the file exists. If it doesn't,
+        -- we can fill in some sensible defaults using the static routing and
+        -- kademlia flags for other nodes.
         topologyOptKademlia <- if nmKademlia
-                                 then liftIO $ Just <$> getKademliaParams cfg
-                                 else return Nothing
+            then do
+                ekparams <- liftIO $ getKademliaParamsFromFile cfg
+                case ekparams of
+                    Right kparams -> return $ Just kparams
+                    Left MissingKademliaConfig ->
+                        let ekparams' = getKademliaParamsFromStatic kademliaPeers
+                        in  either (throw . CannotParseKademliaConfig . Left) (return . Just) ekparams'
+                    Left err -> throw err
+            else return Nothing
         case nmType of
           T.NodeCore  -> return $ T.TopologyCore{..}
           T.NodeRelay -> return $ T.TopologyRelay{topologyMaxSubscrs = nmMaxSubscrs, ..}
-          T.NodeEdge  -> liftIO $ throwM NetworkConfigSelfEdge
+          T.NodeEdge  -> throw NetworkConfigSelfEdge
       Y.TopologyBehindNAT{..} ->
         return T.TopologyBehindNAT{..}
       Y.TopologyP2P{..} -> do
-        kparams <- liftIO $ getKademliaParams cfg
+        ekparams <- liftIO $ getKademliaParamsFromFile cfg
+        kparams <- either throw return ekparams
         return T.TopologyP2P{topologyKademlia = kparams, ..}
       Y.TopologyTraditional{..} -> do
-        kparams <- liftIO $ getKademliaParams cfg
+        ekparams <- liftIO $ getKademliaParamsFromFile cfg
+        kparams <- either throw return ekparams
```

Comment:
I'd usually combine these things into one-liners rather than have to make up an intermediate name, e.g.

    kparams <- either throw return =<< liftIO (getKademliaParamsFromFile cfg)